### DISCUSSION TITLE
Package version number release scheme for packages in the django-cms ecosystem with focus on django-cms 4.x compatibility

### DISCUSSION THREAD

**Discussion by Aiky30** [django-cms/django-cms#7149] [General]
Currently django-cms 3.x compatible packages are released using semver, where a major version requires the package to be incremented e.g. v2.0.0 would become v3.0.0. This complicates django-cms 4.x packages that are also django-cms 3.x packges where we have a branch `support/django-cms-4.0.x`.

We need a scheme to allow us to release django-cms 3.x and 4.x packages from the same repository but in a way that doesn't conflict. 

Possible solutions:
- Use a super high number like 10.x
- Just move to the next version number potentially pausing the issue
- Carry on and continue to use conflicting version numbers X.0.0devX (Not ideal)
- Some other idea?

**Comment by macolo** [comment]
Today's TC participants follow @marksweb's suggestion to coordinate a django CMS v3 sunset roadmap for packages in the periphery of django-cms. 

The following is based on @marksweb 's thoughts: 

## Principles

- The following concerns any package in the django CMS ecosystem with the exception of django-cms itself.
- Anyone can introduce django CMS 4 compatibility to a package, at any time
- For django CMS 4 compatible releases, use the next major version for that release. 
- Upon release, a maintainer should announce limited django CMS v3 support (i.e. minor patches / fixes only). 
- From there onwards the package is considered a "django CMS v4 compatible" package and semver versioning can continue without limitations 
- The previous major version serves to maintain limited django CMS v3 support.

## Example
- django-cms-bootstrap4 has currently version 2.0.0
- Someone introduces django cms 4 compatibility under version 3.0.0
- Someone adds a small fix: new version 3.0.1
- Someone fixes a security issue and backports it: new versions 2.0.1 and 3.0.1

**Comment by Aiky30** [comment]
I need to draw / describe in detail the version numbers and scenarios for the packages to ensure that we have everything covered.

**Reply by macolo** [reply]
@Aiky30 I structured the proposal a bit better, is this good enough for adding to the official documentation?

**Comment by marksweb** [comment]
If we decide on my branching idea for legacy support, I think it might help this if we can figure out what the path to releasing cms 4 is. And alongside that, what other packages need/have that cms 4 support so that they can be considered.

Then once we know what we need to do to get them into a releasable state we can make final cms 3.x supporting versions & then setup the legacy support branches & get to releasing cms 4.

**Comment by Aiky30** [comment]
# A proposal of package versions with django-cms 4 codebases

**THIS ISSUE IS NOW CRITICAL with django 2.2, 3.2 and 4.x support required!!!!!!!!!!**

The proposals take into consideration that packages for django-cms 4 are released in production and still need to support:
    - django 2.2
    - django 3.2, needs to freely allow the previous django 2.x codebases to be patched
    - django 4.x, needs to freely allow the previous django 3.x codebases to be patched

Currently all packages are on a limited dev stream of for example 4.0.0.devx, this doesn't allow for any bugfixes, features or patches if you have major changes between dev releases. 

**A scenario:** A package has django 2.2 support in 0.0.0.dev1, then in 0.0.0.dev2 you drop django 2.2 support, how do you patch any changes to the code inbetween dev1 and dev2? It's a fact that django-cms 4 packages are now stuck at this crossroad with no way of accepting django 3.2 changes never mind django 4.x changes. This is fine in django-cms 3.x codebases because you have whole version numbers and the branching model to accept this, with limited to no area for package version release numbers this issue is going to become a huge mess!

## Proposed branching and release strategy 

### django-cms
#### Current
    - 4.0.0.devx django 2.2
        - Branch: release/4.0.x
    - Not released django 3.x
        - Branch: develop-4
   
#### Proposal until django-cms v4 is official
    - 4.0.0.devx django 2.2
        - Branch: release/4.0.0.x
    - 4.0.1.devx django 3.x
        - Branch: release/4.0.1.x
    - 4.0.2.devx django 4.x
        - Branch: develop-4, future branch: 4.0.3.devx

### djangocms-text-ckeditor
#### Current
    - django-cms 3.x compatible versions released:
        - 4.0.0, 5.0.0, 5.0.1
    - 4.0.0.devx django 2.2
        - Branch: support/4.0.0
    - Not released django 3.x
        - Branch: support/4.0.x

#### Proposal until django-cms v4 is official
    - 4.0.0.devx django 2.2
        - Branch: support/4.0.0
    - 4.0.1.devx django 3.x
        - Branch: support/4.0.1
    - Future working changes (develop for django-cms 4 develop-4 changes)
        - Branch: support/4.0.x

### djangocms-snippet 

#### Current
    - django-cms 3.x compatible versions released:
        - 3.0.0
    - 4.0.0.devx django 2.2
        - Branch: support/django-cms-4.0.x
    - Not released django 3.x
        - No branch currently

#### Proposal until django-cms v4 is official
    - 4.0.0.devx django 2.2
        - Branch: support/django-cms-4.0.0
    - 4.0.1.devx django 3.x
        - Branch: support/django-cms-4.0.1
    - Future working changes (develop for django-cms 4 develop-4 changes)
        - Branch: support/django-cms-4.0.x

**Reply by jrief** [reply]
As discussed on our board meeting, I'm OK with this.

**Reply by fsbraun** [reply]
I agree with this proposal! It should take confusion out and make PR easier!

**Comment by marksweb** [comment]
Consider this signed off 👍

**Comment by Aiky30** [comment]
For full clarity:

I have amended the proposal slightly for django-cms where the current branch was incorrectly referenced.

I have also found that I overlooked the current release/4.0.x branch of the cms. This branch will still be relevant in the future when 4.0 is an official release but with the proposal will be redundant because develop-4 serves the same purpose. I don't see any issues with this and the contents of develop-4 can just be merged directly into release/4.0.x in the future.
